File: ruby/native/HelloWorldKomencoNative/QuantumCircuit.rb
module Automatski
  module Komenco
  
    class Operation
      attr_reader :gate, :qubits, :params

      def initialize(gate, qubits, params)
        @gate = gate
        @qubits = qubits
        @params = params
      end

      def get_gate
        @gate
      end

      def get_qubits
        @qubits
      end

      def get_params
        @params
      end
    end
	
    class QuantumCircuit
      attr_reader :num_qubits, :operations

      def initialize(num_qubits)
        @num_qubits = num_qubits
        @operations = []
      end

      def get_num_qubits
        @num_qubits 
      end
	  
	  def get_operations
        @operations
      end

      def add_single(gate, params, qubits)
        if qubits.size != 1
          raise ArgumentError, "Number of qubits for gate #{gate} has to be one"
        end
        qubits.each do |qubit|
          if qubit < 0 || qubit >= @num_qubits
            raise ArgumentError, "Invalid qubit: #{qubit}"
          end
        end
        @operations << Operation.new(gate, qubits, params)
      end

      def add_double(gate, params, qubits)
        if qubits.size != 2
          raise ArgumentError, "Number of qubits for gate #{gate} has to be two"
        end
        qubits.each do |qubit|
          if qubit < 0 || qubit >= @num_qubits
            raise ArgumentError, "Invalid qubit: #{qubit}"
          end
        end
        @operations << Operation.new(gate, qubits, params)
      end

      def add_triple(gate, params, qubits)
        if qubits.size != 3
          raise ArgumentError, "Number of qubits for gate #{gate} has to be three"
        end
        qubits.each do |qubit|
          if qubit < 0 || qubit >= @num_qubits
            raise ArgumentError, "Invalid qubit: #{qubit}"
          end
        end
        @operations << Operation.new(gate, qubits, params)
      end

      def add_multiple(gate, params, qubits)
        required_qubits = Math.log(gate.length) / Math.log(2)
        if qubits.size != required_qubits.to_i
          raise ArgumentError, "Number of qubits for gate #{gate} has to be #{required_qubits.to_i}"
        end
        qubits.each do |qubit|
          if qubit < 0 || qubit >= @num_qubits
            raise ArgumentError, "Invalid qubit: #{qubit}"
          end
        end
        @operations << Operation.new(gate, qubits, params)
      end

      def id(qubit1)
        add_single("id", [], [qubit1])
      end

      def x(qubit1)
        add_single("x", [], [qubit1])
      end

      def y(qubit1)
        add_single("y", [], [qubit1])
      end

      def z(qubit1)
        add_single("z", [], [qubit1])
      end

      def h(qubit1)
        add_single("h", [], [qubit1])
      end

      def s(qubit1)
        add_single("s", [], [qubit1])
      end

      def sdg(qubit1)
        add_single("sdg", [], [qubit1])
      end

      def t(qubit1)
        add_single("t", [], [qubit1])
      end

      def tdg(qubit1)
        add_single("tdg", [], [qubit1])
      end

      def rx(theta, qubit1)
        add_single("rx", [theta], [qubit1])
      end

      def ry(theta, qubit1)
        add_single("ry", [theta], [qubit1])
      end

      def rz(theta, qubit1)
        add_single("rz", [theta], [qubit1])
      end

      def u(theta, phi, lam, qubit1)
        add_single("u", [theta, phi, lam], [qubit1])
      end

      def u1(theta, qubit1)
        add_single("u1", [theta], [qubit1])
      end

      def u2(phi, lam, qubit1)
        add_single("u2", [phi, lam], [qubit1])
      end

      def u3(theta, phi, lam, qubit1)
        add_single("u3", [theta, phi, lam], [qubit1])
      end

      def sx(qubit1)
        add_single("sx", [], [qubit1])
      end

      def sxdg(qubit1)
        add_single("sxdg", [], [qubit1])
      end

      def r(theta, phi, qubit1)
        add_single("r", [theta, phi], [qubit1])
      end

      def cx(qubit1, qubit2)
        add_double("cx", [], [qubit1, qubit2])
      end

      def cy(qubit1, qubit2)
        add_double("cy", [], [qubit1, qubit2])
      end

      def cz(qubit1, qubit2)
        add_double("cz", [], [qubit1, qubit2])
      end

      def ch(qubit1, qubit2)
        add_double("ch", [], [qubit1, qubit2])
      end

      def ucrx(theta, qubit1, qubit2)
        add_double("ucrx", [theta], [qubit1, qubit2])
      end

      def ucry(theta, qubit1, qubit2)
        add_double("ucry", [theta], [qubit1, qubit2])
      end

      def ucrz(theta, qubit1, qubit2)
        add_double("ucrz", [theta], [qubit1, qubit2])
      end

      def crx(theta, qubit1, qubit2)
        add_double("crx", [theta], [qubit1, qubit2])
      end

      def cry(theta, qubit1, qubit2)
        add_double("cry", [theta], [qubit1, qubit2])
      end

      def crz(theta, qubit1, qubit2)
        add_double("crz", [theta], [qubit1, qubit2])
      end

      def cr(theta, phi, lam, qubit1, qubit2)
        add_double("cr", [theta, phi, lam], [qubit1, qubit2])
      end

      def cu(theta, phi, lam, gamma, qubit1, qubit2)
        add_double("cu1", [theta, phi, lam, gamma], [qubit1, qubit2])
      end

      def cu1(theta, qubit1, qubit2)
        add_double("cu1", [theta], [qubit1, qubit2])
      end

      def cu2(phi, lam, qubit1, qubit2)
        add_double("cu2", [phi, lam], [qubit1, qubit2])
      end

      def cu3(theta, phi, lam, qubit1, qubit2)
        add_double("cu3", [theta, phi, lam], [qubit1, qubit2])
      end

      def dcx(qubit1, qubit2)
        add_double("dcx", [], [qubit1, qubit2])
      end

      def ecr(qubit1, qubit2)
        add_double("ecr", [], [qubit1, qubit2])
      end

      def iswap(qubit1, qubit2)
        add_double("iswap", [], [qubit1, qubit2])
      end

      def rxx(theta, qubit1, qubit2)
        add_double("rxx", [theta], [qubit1, qubit2])
      end

      def ryy(theta, qubit1, qubit2)
        add_double("ryy", [theta], [qubit1, qubit2])
      end

      def rzz(theta, qubit1, qubit2)
        add_double("rzz", [theta], [qubit1, qubit2])
      end

      def rzx(theta, qubit1, qubit2)
        add_double("rzx", [theta], [qubit1, qubit2])
      end

      def cswap(qubit1, qubit2, qubit3)
        add_triple("cswap", [], [qubit1, qubit2, qubit3])
      end

      def ccx(qubit1, qubit2, qubit3)
        add_triple("ccx", [], [qubit1, qubit2, qubit3])
      end

      def ccy(qubit1, qubit2, qubit3)
        add_triple("ccy", [], [qubit1, qubit2, qubit3])
      end

      def ccz(qubit1, qubit2, qubit3)
        add_triple("ccz", [], [qubit1, qubit2, qubit3])
      end

      def c3x(qubits)
        add_multiple("c3x", [], qubits)
      end

      def c4x(qubits)
        add_multiple("c4x", [], qubits)
      end

      def mcx(qubits)
        add_multiple("mcx", [], qubits)
      end

      def mcu1(theta, qubits)
        add_multiple("mcu1", [theta], qubits)
      end

      def mcu2(phi, lam, qubits)
        add_multiple("mcu2", [phi, lam], qubits)
      end

      def mcu3(theta, phi, lam, qubits)
        add_multiple("mcu3", [theta, phi, lam], qubits)
      end

      def mcz(qubits)
        add_multiple("mcz", [], qubits)
      end

      def mcp(qubits)
        add_multiple("mcp", [], qubits)
      end

      def mcrx(qubits)
        add_multiple("mcrx", [], qubits)
      end

      def mcry(qubits)
        add_multiple("mcry", [], qubits)
      end

      def mcrz(qubits)
        add_multiple("mcrz", [], qubits)
      end

      def qft(qubits)
        add_multiple("qft", [], qubits)
      end

      def inverse_qft(qubits)
        add_multiple("iqft", [], qubits)
      end

      def swap(qubit1, qubit2)
        add_double("swap", [], [qubit1, qubit2])
      end

      def csx(qubit1, qubit2)
        add_double("csx", [], [qubit1, qubit2])
      end

      def p(theta, qubit1)
        add_single("p", [theta], [qubit1])
      end

      def cp(theta, qubit1, qubit2)
        add_double("cp", [theta], [qubit1, qubit2])
      end

      def ccp(theta, qubit1, qubit2, qubit3)
        add_triple("ccp", [theta], [qubit1, qubit2, qubit3])
      end

      def sqrt_x(qubit1)
        add_single("sqrt_x", [], [qubit1])
      end

      def sqrt_y(qubit1)
        add_single("sqrt_y", [], [qubit1])
      end

      def sqrt_z(qubit1)
        add_single("sqrt_z", [], [qubit1])
      end

      def gpi(phi, qubit1)
        add_single("gpi", [phi], [qubit1])
      end

      def gpi2(phi, qubit1)
        add_single("gpi2", [phi], [qubit1])
      end

      def xp(theta, qubit1)
        add_single("xp", [theta], [qubit1])
      end

      def yp(theta, qubit1)
        add_single("yp", [theta], [qubit1])
      end

      def zp(theta, qubit1)
        add_single("zp", [theta], [qubit1])
      end

      def xxp(theta, qubit1, qubit2)
        add_double("xxp", [theta], [qubit1, qubit2])
      end

      def yyp(theta, qubit1, qubit2)
        add_double("yyp", [theta], [qubit1, qubit2])
      end

      def zzp(theta, qubit1, qubit2)
        add_double("zzp", [theta], [qubit1, qubit2])
      end

      def phased_xp(theta, phi, qubit1)
        add_single("phased_xp", [theta, phi], [qubit1])
      end

      def phased_yp(theta, phi, qubit1)
        add_single("phased_yp", [theta, phi], [qubit1])
      end

      def phased_zp(theta, phi, qubit1)
        add_single("phased_zp", [theta, phi], [qubit1])
      end

      def cnotp(theta, qubit1, qubit2)
        add_double("cnotp", [theta], [qubit1, qubit2])
      end

      def ccnotp(theta, qubit1, qubit2, qubit3)
        add_triple("ccnotp", [theta], [qubit1, qubit2, qubit3])
      end

      def cyp(theta, qubit1, qubit2)
        add_double("cyp", [theta], [qubit1, qubit2])
      end

      def ccyp(theta, qubit1, qubit2, qubit3)
        add_triple("ccyp", [theta], [qubit1, qubit2, qubit3])
      end

      def czp(theta, qubit1, qubit2)
        add_double("czp", [theta], [qubit1, qubit2])
      end

      def cczp(theta, qubit1, qubit2, qubit3)
        add_triple("cczp", [theta], [qubit1, qubit2, qubit3])
      end

      def measure(qubits)
        cparams = []
        cqubits = []
        cgate = "measure"
        if qubits.empty?
          raise ArgumentError, "number of qubits to measure cannot be zero"
        else
          qubits.each do |qubit|
            if qubit < 0 || qubit >= @num_qubits
              raise ArgumentError, "invalid qubit: #{qubit}"
            else
              cqubits << qubit
            end
          end
        end
        @operations << Operation.new(cgate, cqubits, cparams)
      end

      def measure_all
        cparams = []
        cqubits = []
        cgate = "measure"
        (0...@num_qubits).each do |qubit|
          cqubits << qubit
        end
        @operations << Operation.new(cgate, cqubits, cparams)
      end

      def random_circuit(number_of_operations, qubits_to_use_for_random_circuit)
        qubits_count = qubits_to_use_for_random_circuit.size
        if qubits_count < 3
          raise ArgumentError, "The number of qubits needed to generate random circuits is >= 3"
        end

        gates_and_qubits_map = {
          # 1 qubit gates
          "x" => 1, "y" => 1, "z" => 1, "rx" => 1, "ry" => 1, "rz" => 1,
          "h" => 1, "s" => 1, "sdg" => 1, "t" => 1, "tdg" => 1, "sx" => 1, "sxdg" => 1,
          # 2 qubit gates
          "cx" => 2, "cy" => 2, "cz" => 2, "crx" => 2, "cry" => 2, "crz" => 2,
          "ch" => 2, "dcx" => 2, "ecr" => 2, "iswap" => 2, "swap" => 2, "csx" => 2,
          # 3 qubit gates
          "cswap" => 3, "ccx" => 3
        }

        all_possible_gates = gates_and_qubits_map.keys
        all_possible_gates_count = all_possible_gates.size

        number_of_operations.times do
          index = rand(all_possible_gates_count)
          cgate = all_possible_gates[index]

          number_of_qubits = gates_and_qubits_map[cgate]

          qubits_to_use_for_random_circuit.shuffle!
          cqubits = qubits_to_use_for_random_circuit.take(number_of_qubits)

          cparams = []
          if ["rx", "ry", "rz", "crx", "cry", "crz"].include?(cgate)
            theta = rand * 2 * Math::PI
            cparams << theta
          end

          @operations << Operation.new(cgate, cqubits, cparams)
        end
      end
    end
	
  end
end

File: dart/native/helloworldkomenconative/bin/quantumcircuit.dart
import 'dart:math';

class Operation {
  final String gate;
  final List<int> qubits;
  final List<double> params;

  Operation(this.gate, this.qubits, this.params);

  String getGate() {
    return gate;
  }

  List<int> getQubits() {
    return qubits;
  }

  List<double> getParams() {
    return params;
  }
}

class QuantumCircuit {
  final int numQubits;
  final List<Operation> operations = [];

  QuantumCircuit(this.numQubits);

  int getNumQubits() {
    return numQubits;
  }

  List<Operation> getOperations() {
    return operations;
  }

  void _addSingle(String gate, List<double> params, List<int> qubits) {
    if (qubits.length != 1) {
      throw ArgumentError("Number of qubits for gate $gate has to be one");
    }
    for (var qubit in qubits) {
      if (qubit < 0 || qubit >= numQubits) {
        throw ArgumentError("Invalid qubit: $qubit");
      }
    }
    operations.add(Operation(gate, qubits, params));
  }

  void _addDouble(String gate, List<double> params, List<int> qubits) {
    if (qubits.length != 2) {
      throw ArgumentError("Number of qubits for gate $gate has to be two");
    }
    for (var qubit in qubits) {
      if (qubit < 0 || qubit >= numQubits) {
        throw ArgumentError("Invalid qubit: $qubit");
      }
    }
    operations.add(Operation(gate, qubits, params));
  }

  void _addTriple(String gate, List<double> params, List<int> qubits) {
    if (qubits.length != 3) {
      throw ArgumentError("Number of qubits for gate $gate has to be three");
    }
    for (var qubit in qubits) {
      if (qubit < 0 || qubit >= numQubits) {
        throw ArgumentError("Invalid qubit: $qubit");
      }
    }
    operations.add(Operation(gate, qubits, params));
  }

  void _addMultiple(String gate, List<double> params, List<int> qubits) {
    int requiredQubits = (log(gate.length) / log(2)).toInt();
    if (qubits.length != requiredQubits) {
      throw ArgumentError("Number of qubits for gate $gate has to be $requiredQubits");
    }
    for (var qubit in qubits) {
      if (qubit < 0 || qubit >= numQubits) {
        throw ArgumentError("Invalid qubit: $qubit");
      }
    }
    operations.add(Operation(gate, qubits, params));
  }

  void id(int qubit1) {
    _addSingle("id", [], [qubit1]);
  }

  void x(int qubit1) {
    _addSingle("x", [], [qubit1]);
  }

  void y(int qubit1) {
    _addSingle("y", [], [qubit1]);
  }

  void z(int qubit1) {
    _addSingle("z", [], [qubit1]);
  }

  void h(int qubit1) {
    _addSingle("h", [], [qubit1]);
  }

  void s(int qubit1) {
    _addSingle("s", [], [qubit1]);
  }

  void sdg(int qubit1) {
    _addSingle("sdg", [], [qubit1]);
  }

  void t(int qubit1) {
    _addSingle("t", [], [qubit1]);
  }

  void tdg(int qubit1) {
    _addSingle("tdg", [], [qubit1]);
  }

  void rx(double theta, int qubit1) {
    _addSingle("rx", [theta], [qubit1]);
  }

  void ry(double theta, int qubit1) {
    _addSingle("ry", [theta], [qubit1]);
  }

  void rz(double theta, int qubit1) {
    _addSingle("rz", [theta], [qubit1]);
  }

  void u(double theta, double phi, double lam, int qubit1) {
    _addSingle("u", [theta, phi, lam], [qubit1]);
  }

  void u1(double theta, int qubit1) {
    _addSingle("u1", [theta], [qubit1]);
  }

  void u2(double phi, double lam, int qubit1) {
    _addSingle("u2", [phi, lam], [qubit1]);
  }

  void u3(double theta, double phi, double lam, int qubit1) {
    _addSingle("u3", [theta, phi, lam], [qubit1]);
  }

  void sx(int qubit1) {
    _addSingle("sx", [], [qubit1]);
  }

  void sxdg(int qubit1) {
    _addSingle("sxdg", [], [qubit1]);
  }

  void r(double theta, double phi, int qubit1) {
    _addSingle("r", [theta, phi], [qubit1]);
  }

  void cx(int qubit1, int qubit2) {
    _addDouble("cx", [], [qubit1, qubit2]);
  }

  void cy(int qubit1, int qubit2) {
    _addDouble("cy", [], [qubit1, qubit2]);
  }

  void cz(int qubit1, int qubit2) {
    _addDouble("cz", [], [qubit1, qubit2]);
  }

  void ch(int qubit1, int qubit2) {
    _addDouble("ch", [], [qubit1, qubit2]);
  }

  void ucrx(double theta, int qubit1, int qubit2) {
    _addDouble("ucrx", [theta], [qubit1, qubit2]);
  }

  void ucry(double theta, int qubit1, int qubit2) {
    _addDouble("ucry", [theta], [qubit1, qubit2]);
  }

  void ucrz(double theta, int qubit1, int qubit2) {
    _addDouble("ucrz", [theta], [qubit1, qubit2]);
  }

  void crx(double theta, int qubit1, int qubit2) {
    _addDouble("crx", [theta], [qubit1, qubit2]);
  }

  void cry(double theta, int qubit1, int qubit2) {
    _addDouble("cry", [theta], [qubit1, qubit2]);
  }

  void crz(double theta, int qubit1, int qubit2) {
    _addDouble("crz", [theta], [qubit1, qubit2]);
  }

  void cr(double theta, double phi, double lam, int qubit1, int qubit2) {
    _addDouble("cr", [theta, phi, lam], [qubit1, qubit2]);
  }

  void cu(double theta, double phi, double lam, double gamma, int qubit1, int qubit2) {
    _addDouble("cu1", [theta, phi, lam, gamma], [qubit1, qubit2]);
  }

  void cu1(double theta, int qubit1, int qubit2) {
    _addDouble("cu1", [theta], [qubit1, qubit2]);
  }

  void cu2(double phi, double lam, int qubit1, int qubit2) {
    _addDouble("cu2", [phi, lam], [qubit1, qubit2]);
  }

  void cu3(double theta, double phi, double lam, int qubit1, int qubit2) {
    _addDouble("cu3", [theta, phi, lam], [qubit1, qubit2]);
  }

  void dcx(int qubit1, int qubit2) {
    _addDouble("dcx", [], [qubit1, qubit2]);
  }

  void ecr(int qubit1, int qubit2) {
    _addDouble("ecr", [], [qubit1, qubit2]);
  }

  void iswap(int qubit1, int qubit2) {
    _addDouble("iswap", [], [qubit1, qubit2]);
  }

  void rxx(double theta, int qubit1, int qubit2) {
    _addDouble("rxx", [theta], [qubit1, qubit2]);
  }

  void ryy(double theta, int qubit1, int qubit2) {
    _addDouble("ryy", [theta], [qubit1, qubit2]);
  }

  void rzz(double theta, int qubit1, int qubit2) {
    _addDouble("rzz", [theta], [qubit1, qubit2]);
  }

  void rzx(double theta, int qubit1, int qubit2) {
    _addDouble("rzx", [theta], [qubit1, qubit2]);
  }

  void cswap(int qubit1, int qubit2, int qubit3) {
    _addTriple("cswap", [], [qubit1, qubit2, qubit3]);
  }

  void ccx(int qubit1, int qubit2, int qubit3) {
    _addTriple("ccx", [], [qubit1, qubit2, qubit3]);
  }

  void ccy(int qubit1, int qubit2, int qubit3) {
    _addTriple("ccy", [], [qubit1, qubit2, qubit3]);
  }

  void ccz(int qubit1, int qubit2, int qubit3) {
    _addTriple("ccz", [], [qubit1, qubit2, qubit3]);
  }

  void c3x(List<int> qubits) {
    _addMultiple("c3x", [], qubits);
  }

  void c4x(List<int> qubits) {
    _addMultiple("c4x", [], qubits);
  }

  void mcx(List<int> qubits) {
    _addMultiple("mcx", [], qubits);
  }

  void mcu1(double theta, List<int> qubits) {
    _addMultiple("mcu1", [theta], qubits);
  }

  void mcu2(double phi, double lam, List<int> qubits) {
    _addMultiple("mcu2", [phi, lam], qubits);
  }

  void mcu3(double theta, double phi, double lam, List<int> qubits) {
    _addMultiple("mcu3", [theta, phi, lam], qubits);
  }

  void mcz(List<int> qubits) {
    _addMultiple("mcz", [], qubits);
  }

  void mcp(List<int> qubits) {
    _addMultiple("mcp", [], qubits);
  }

  void mcrx(List<int> qubits) {
    _addMultiple("mcrx", [], qubits);
  }

  void mcry(List<int> qubits) {
    _addMultiple("mcry", [], qubits);
  }

  void mcrz(List<int> qubits) {
    _addMultiple("mcrz", [], qubits);
  }

  void qft(List<int> qubits) {
    _addMultiple("qft", [], qubits);
  }

  void inverseQft(List<int> qubits) {
    _addMultiple("iqft", [], qubits);
  }

  void swap(int qubit1, int qubit2) {
    _addDouble("swap", [], [qubit1, qubit2]);
  }

  void csx(int qubit1, int qubit2) {
    _addDouble("csx", [], [qubit1, qubit2]);
  }

  void p(double theta, int qubit1) {
    _addSingle("p", [theta], [qubit1]);
  }

  void cp(double theta, int qubit1, int qubit2) {
    _addDouble("cp", [theta], [qubit1, qubit2]);
  }

  void ccp(double theta, int qubit1, int qubit2, int qubit3) {
    _addTriple("ccp", [theta], [qubit1, qubit2, qubit3]);
  }

  void sqrtX(int qubit1) {
    _addSingle("sqrt_x", [], [qubit1]);
  }

  void sqrtY(int qubit1) {
    _addSingle("sqrt_y", [], [qubit1]);
  }

  void sqrtZ(int qubit1) {
    _addSingle("sqrt_z", [], [qubit1]);
  }

  void gpi(double phi, int qubit1) {
    _addSingle("gpi", [phi], [qubit1]);
  }

  void gpi2(double phi, int qubit1) {
    _addSingle("gpi2", [phi], [qubit1]);
  }

  void xp(double theta, int qubit1) {
    _addSingle("xp", [theta], [qubit1]);
  }

  void yp(double theta, int qubit1) {
    _addSingle("yp", [theta], [qubit1]);
  }

  void zp(double theta, int qubit1) {
    _addSingle("zp", [theta], [qubit1]);
  }

  void xxp(double theta, int qubit1, int qubit2) {
    _addDouble("xxp", [theta], [qubit1, qubit2]);
  }

  void yyp(double theta, int qubit1, int qubit2) {
    _addDouble("yyp", [theta], [qubit1, qubit2]);
  }

  void zzp(double theta, int qubit1, int qubit2) {
    _addDouble("zzp", [theta], [qubit1, qubit2]);
  }

  void phasedXp(double theta, double phi, int qubit1) {
    _addSingle("phased_xp", [theta, phi], [qubit1]);
  }

  void phasedYp(double theta, double phi, int qubit1) {
    _addSingle("phased_yp", [theta, phi], [qubit1]);
  }

  void phasedZp(double theta, double phi, int qubit1) {
    _addSingle("phased_zp", [theta, phi], [qubit1]);
  }

  void cnotp(double theta, int qubit1, int qubit2) {
    _addDouble("cnotp", [theta], [qubit1, qubit2]);
  }

  void ccnotp(double theta, int qubit1, int qubit2, int qubit3) {
    _addTriple("ccnotp", [theta], [qubit1, qubit2, qubit3]);
  }

  void cyp(double theta, int qubit1, int qubit2) {
    _addDouble("cyp", [theta], [qubit1, qubit2]);
  }

  void ccyp(double theta, int qubit1, int qubit2, int qubit3) {
    _addTriple("ccyp", [theta], [qubit1, qubit2, qubit3]);
  }

  void czp(double theta, int qubit1, int qubit2) {
    _addDouble("czp", [theta], [qubit1, qubit2]);
  }

  void cczp(double theta, int qubit1, int qubit2, int qubit3) {
    _addTriple("cczp", [theta], [qubit1, qubit2, qubit3]);
  }

  void measure(List<int> qubits) {
    List<double> cparams = [];
    List<int> cqubits = [];
    String cgate = "measure";
    if (qubits.isEmpty) {
      throw ArgumentError("number of qubits to measure cannot be zero");
    } else {
      for (var qubit in qubits) {
        if (qubit < 0 || qubit >= numQubits) {
          throw ArgumentError("invalid qubit: $qubit");
        } else {
          cqubits.add(qubit);
        }
      }
    }
    operations.add(Operation(cgate, cqubits, cparams));
  }

  void measureAll() {
    List<double> cparams = [];
    List<int> cqubits = [];
    String cgate = "measure";
    for (var qubit = 0; qubit < numQubits; qubit++) {
      cqubits.add(qubit);
    }
    operations.add(Operation(cgate, cqubits, cparams));
  }

  void randomCircuit(int numberOfOperations, List<int> qubitsToUseForRandomCircuit) {
    int qubitsCount = qubitsToUseForRandomCircuit.length;
    if (qubitsCount < 3) {
      throw ArgumentError("The number of qubits needed to generate random circuits is >= 3");
    }

    Map<String, int> gatesAndQubitsMap = {
      // 1 qubit gates
      "x": 1, "y": 1, "z": 1, "rx": 1, "ry": 1, "rz": 1,
      "h": 1, "s": 1, "sdg": 1, "t": 1, "tdg": 1, "sx": 1, "sxdg": 1,
      // 2 qubit gates
      "cx": 2, "cy": 2, "cz": 2, "crx": 2, "cry": 2, "crz": 2,
      "ch": 2, "dcx": 2, "ecr": 2, "iswap": 2, "swap": 2, "csx": 2,
      // 3 qubit gates
      "cswap": 3, "ccx": 3
    };

    List<String> allPossibleGates = gatesAndQubitsMap.keys.toList();
    int allPossibleGatesCount = allPossibleGates.length;
    Random random = Random();

    for (int i = 0; i < numberOfOperations; i++) {
      int index = random.nextInt(allPossibleGatesCount);
      String cgate = allPossibleGates[index];

      int numberOfQubits = gatesAndQubitsMap[cgate]!;

      qubitsToUseForRandomCircuit.shuffle(random);
      List<int> cqubits = qubitsToUseForRandomCircuit.sublist(0, numberOfQubits);

      List<double> cparams = [];
      if (["rx", "ry", "rz", "crx", "cry", "crz"].contains(cgate)) {
        double theta = random.nextDouble() * 2 * pi;
        cparams.add(theta);
      }

      operations.add(Operation(cgate, cqubits, cparams));
    }
  }
}